src/language-js/needs-parens.js

```diff
@@ -801,6 +807,63 @@ function isStatement(node) {
   );
 }
 
+function includesFunctionTypeInObjectType(path) {
```

Comment:
There exists a `hasNode` function In `language-js/utils.js`. Please reuse it instead of introducing another function for recursive traversing of the AST.

Reply:
I tried to implement `includesFunctionTypeInObjectType` with `hasNode` like below:

```js
function includesFunctionTypeInObjectType(path) {
  return hasNode(path.getValue(), node => {
    switch (node.type) {
      case "FunctionTypeAnnotation": {
        return true;
      }
      case undefined:
        return false;
      default:
        return node;
    }
  });
}
```

However, run test them failed in following case:

```js
// Input(expected output)
const f = (): a | (string => string) => {};
const f = (): a & (string => string) => {};

// Output
const f = (): (a | (string => string)) => {};
const f = (): (a & (string => string)) => {};
```

To avoid this, it is necessary to refer to the ancestor of the node in `case" FunctionTypeAnnotation`. But since the `path` cannot be used in `hasNode`, it is not possible to go back to the ancestor.

So, at least as far as I did, the way to use hasNode cannot work...

Reply:
Use `hasNode` inside `hasNode` :)

Reply:
Could you suggest? I have no idea...

Reply:
```js
  case "TypeAnnotation":
    return (
      name === "returnType" &&
      parent.type === "ArrowFunctionExpression" &&
      hasNode(
        node,
        n1 =>
          (n1.type === "ObjectTypeAnnotation" &&
            hasNode(
              n1,
              n2 => n2.type === "FunctionTypeAnnotation" || undefined
            )) ||
          undefined
      )
    );
```

Reply:
It looks great! thanks.